node/src/Pos/Block/Network/Retrieval.hs

```diff
@@ -314,7 +318,11 @@ handleCHsValid enqueue endedRecoveryVar nodeId lcaChild newestHash = do
                                 then isJust <$> tryTakeTMVar recHeaderVar
                                 else return False
                 void $ tryPutMVar endedRecoveryVar endedRecovery
-    void $ waitForConversations its
+    -- Try to fill the MVar at the end. In case the conversation is run
+    -- and succesfully ends it, it will already be filled. That's fine.
+    -- But in case there's an error retrieving the blocks, or in case the
+    -- conversation is enqueued to nobody, we have to ensure the MVar is filled.
+    void $ waitForConversations its `finally` tryPutMVar endedRecoveryVar False
```

Comment:
And I see now that this problem is present even without this patch. It's always been assumed that the conversation body will run at most once.